src/main/java/kitchenpos/menu/domain/MenuValidator.java

```diff
@@ -0,0 +1,81 @@
+package kitchenpos.menu.domain;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import kitchenpos.common.vo.Price;
+import kitchenpos.menu.domain.repository.MenuGroupRepository;
+import kitchenpos.product.domain.Product;
+import kitchenpos.product.domain.ProductRepository;
+import org.springframework.stereotype.Component;
+
+@Component
+public class MenuValidator {
+
+    private final ProductRepository productRepository;
+    private final MenuGroupRepository menuGroupRepository;
+
+    public MenuValidator(final ProductRepository productRepository, final MenuGroupRepository menuGroupRepository) {
+        this.productRepository = productRepository;
+        this.menuGroupRepository = menuGroupRepository;
+    }
+
+    public void validateInit(final Menu menu) {
+        validateInvalidMenuGroup(menu);
+        validateInvalidProduct(menu);
+        validateInvalidPrice(menu);
+    }
+
+    private void validateInvalidMenuGroup(final Menu menu) {
+        if (!menuGroupRepository.existsById(menu.getMenuGroupId())) {
+            throw new IllegalArgumentException("존재하지 않는 메뉴 그룹입니다.");
+        }
+    }
+
+    private void validateInvalidProduct(final Menu menu) {
+        final List<Long> requiredProductIds = extractProductIds(menu.getMenuProducts());
+        final Set<Long> savedProducts = findAllProductsByIds(requiredProductIds).stream()
+                .map(Product::getId)
+                .collect(Collectors.toSet());
+        for (Long requiredProductId : requiredProductIds) {
+            if (!savedProducts.contains(requiredProductId)) {
+                throw new IllegalArgumentException("존재하지 않는 상품이 포함되어 있습니다.");
+            }
+        }
```

Comment:
`findAllProductsByIds()`를 통해 조회해 온 Product들을 아이디로 다시 바꾸지 않고 크기만 비교해도 괜찮지 않을까요?
아래와 같이 수정할 수 있을 것 같습니다.

```suggestion
        final List<Long> requiredProductIds = extractProductIds(menu.getMenuProducts());
        final List<Product> savedProducts = findAllProductsByIds(requiredProductIds);
        if (savedProducts.size() != requiredProductIds.size()) {
            throw new IllegalArgumentException("존재하지 않는 상품이 포함되어 있습니다.");
        }
```

Reply:
오호 천재인가요~